File: e14.pl
#!/usr/bin/perl

use strict; 
use warnings;

my $final = 0; 
my $final_chain = 0; 
my $tmp = 0; 
my $len = 0; 

for ( my $i = 1; $i <= 1000000; $i++ ) {
    $len = get_collatz_chain($i); 
    if ( $len > $tmp ) {
        $tmp = $len; 
        $final = $i; 
        $final_chain = $len; 
    }
}
print "$final $final_chain\n"; 
print get_collatz_chain(13), "\n"; 

sub get_collatz_chain {
    my $n = shift;
    my @chain_nodes; 
    push @chain_nodes, $n; 
    while ( $n != 1 ) {
        if ( $n % 2 == 0 ) {
            push @chain_nodes, $n; 
            $n = $n / 2;
        } else {
            $n = ( $n * 3 ) + 1; 
        }
    }
    push @chain_nodes, 1;
    my $length = $#chain_nodes; 
    return $length; 
}

File: e14.c
#include <stdio.h>
#include <stdlib.h>

long get_collatz_chain( long n );

int main() 
{
    long final = 0;
    long tmp = 0;
    long len = 0; 

    long i; 
    for (i = 2; i < 1000001; i++) {
        len = get_collatz_chain(i); 
        if ( len > tmp ) { 
            tmp = len; 
            final = i;
        }
    }
    printf("%ld\n", final); 
    return 0; 
}

long get_collatz_chain( long n ) 
{
    long count = 0; 
    while (n != 1) {
        if (n%2 == 0) {
            n = n/2;
        } else {
            n = (n*3) + 1;
        }
        count++;
    }
    return count;
}

